### PR TITLE
Unified threshold test/metrics > tests

### PR DESCRIPTION
**Pull request by validbeck**
## Internal Notes for Reviewers

> For sc-4829, I either removed or replaced threshold tests & metrics with simply tests whenever possible.

### Assess compliance

| Old | New |
|---|---|
| Screenshot 2024-08-23 at 1 38 47 PM | Screenshot 2024-08-23 at 1 36 14 PM|

### Training for model developers
- These steps were slightly inaccurate on top of referring to threshold tests, so I adjusted them

| Old | New |
|---|---|
| Screenshot 2024-08-23 at 1 37 37 PM | Screenshot 2024-08-23 at 1 37 22 PM|

## Stuff not removed
- If it referred to a quality, performance, or monitoring metric (i.e. a "conceptual" metric, not a test metric) I left it in
- Or, if it referred to a specific python function we talk about that still existed in the developer reference `metric_id` I also left it alone
- If it linked out to a notebook talking about metrics that was still hanging around in the main `developer-framework` branch I also let it be

**Examples:**

| Conceptual metric | Python function | Notebook |
|---|---|---|
| Screenshot 2024-08-23 at 1 33 33 PM | Screenshot 2024-08-23 at 1 09 49 PM | Screenshot 2024-08-23 at 1 12 18 PM |

### Glossary
- I left in this definition because it covers us merging the too concepts, just for clarification

Screenshot 2024-08-23 at 1 32 48 PM

> Below are the search results for `metric` after I combed through them, let me know if any of these can be adjusted or retooled as well. Most of these are release notes, so not too big of a deal. 

```
site/about/overview-model-documentation.qmd:
  113  - Metadata about datasets and models, used to lookup programmatic documentation content, such as the stored definition for _common logistic regression limitations_ when a logistic regression model has been passed to the ValidMind test plan to be run.
  114: - Quality and performance metrics collected from datasets and models.
  115  - Output from test and test suites that have been run.
  116: - Images, plots, visuals that were generated as part of extracting metrics and running tests.
  117  

site/about/fine-print/data-privacy-policy.qmd:
  44  - Metadata about datasets and models, used to look up programmatic documentation content, such as the stored definition for _common logistic regression limitations_ when a logistic regression model has been passed to the ValidMind test suite to be run.
  45: - Quality and performance metrics collected from datasets and models.
  46  - Output from tests and test suites that have been run.
  47: - Images, plots, visuals that were generated as part of extracting metrics and running tests.
  48  

site/about/glossary/key_concepts/_docs.qmd:
  1  model documentation
  2: : A structured and detailed record pertaining to a model, encompassing key components such as its underlying assumptions, methodologies, data sources, inputs, performance metrics, evaluations, limitations, and intended uses. 
  3      

site/about/glossary/key_concepts/_monitoring-template.qmd:
  1  monitoring template
  2: : Functions as a framework for structuring the monitoring of model performance, segmented into various sections and sub-sections. Monitoring templates define the structure of your model performance evaluations, specifying the metrics to be tracked, the tests to be conducted, and how the results should be displayed.
  3  

site/faq/faq-inventory.qmd:
  19  
  20: ValidMind is working on a dashboard feature that provides executive metrics, such as model documentation compliance reporting across BUs, findings by status and model use case, and more.
  21  
  22: These metrics can be exported into a customizable report for the customer.
  23  

site/faq/faq-privacy.qmd:
  18  - Dataset and model metadata which allow generating documentation snippets programmatically (example: stored definition for "common logistic regression limitations" when a logistic regression model has been passed to the ValidMind test suite execution)
  19: - Quality and performance metrics collected from the dataset and model
  20  - Outputs from executed test suites
  21: - Images, plots, and visuals generated as part of extracting metrics and running tests
  22  

  42  
  43: ## Can the tool automatically document other non-standard ETL steps or performance metrics from notebooks?
  44  

  56  
  57: Our {{< var validmind.developer >}} can extract dataset quality metrics on Pandas DataFrame, NumPy arrays, or Spark DataFrame instances using standard metrics provided by popular open-source frameworks such as scikit-learn, statsmodels, and more.
  58  
  59: Each test defines a mapping to the different supported dataset and/or model interfaces: when passing a Spark DataFrame, our {{< var vm.developer >}} will directly call native evaluation metrics provided by the SparkML API or custom ones built by the developer (such as via UDFs).

site/faq/faq-testing.qmd:
  29  - Data evaluation
  30:     - Data quality metrics
  31      - Sampling method validation

site/guide/model-documentation/customize-documentation-templates.qmd:
  116  
  117: ⟪ 34176 characters skipped ⟫num> <h4>Must be one of:</h4> <ul class=list-group><li class="list-group-item enum-item">"metadata_text"</li><li class="list-group-item enum-item">"dynamic"</li><li class="list-group-item enum-item">"metric"</li><li class="list-group-item enum-item">"test"</li></ul> </div> <br> <div class="badge badge-secondary">Examples:</div> <br><div id=sections_items_contents_items_content_type_ex1 class="jumbotron examples"><div class=highlight><pre><span></span><span class=s2>&quot;metadata_text&quot;</span>
  118  </pre></div> </div><div id=sections_items_contents_items_content_type_ex2 class="jumbotron examples"><div class=highlight><pre><span></span><span class=s2>&quot;test&quot;</span>

  124  </pre></div> </div><div id=sections_items_contents_items_options_ex2 class="jumbotron examples"><div class=highlight><pre><span></span><span class=p>{</span>
  125: <span class=w>    </span><span class=s2>&quot;metric_id&quot;</span><span class=o>:</span><span class=w> </span><span class=s2>&quot;metric_1&quot;</span><span class=p>,</span>
  126: <span class=w>    </span><span class=s2>&quot;title&quot;</span><span class=o>:</span><span class=w> </span><span class=s2>&quot;Custom Title for Metric 1&quot;</span>
  127  <span class=p>}</span>

site/guide/monitoring/ongoing-monitoring.qmd:
   58  
   59: As monitoring continues, it's important to report model status to key stakeholders, such as your model risk management committee. Regular updates with summary metrics will keep stakeholder informed of findings and emerging risks, highlighting significant trends or issues that may need action.
   60  

   62  
   63: - **Metrics and models** — Specify which models and performance metrics will be monitored.
   64: - **Monitoring frequency** — Set how often each metric will be monitored, based on the model's risk and importance.
   65: - **Risk thresholds** — Establish thresholds that trigger alerts or actions when metrics deviate from expected ranges.
   66  - **Notification system** — Implement a system to notify stakeholders when significant issues or deviations occur.

  101  
  102: The design of your ongoing monitoring plan overall should be a collaborative effort between the first line of defense, typically business units or model owners, and the second line of defense, namely your risk management team. Together, they should select the performance metrics, determine monitoring frequency, and tailor the ongoing monitoring plan to the model's specific use case and risk profile.
  103  

site/releases/2023/2023-aug-15/highlights.qmd:
  14  
  15: <!---John6797/sc 1980/metric test class dx improvements by @johnwalz97 in [#220](https://github.com/validmind/validmind-python/pull/220)--->
  16: - **Improved developer experience for the Metric test class**. We made a number of enhancements to the {{< var validmind.developer >}} to improve the experience for developers:
  17  

site/releases/2023/2023-nov-09/highlights.qmd:
  31  <!---- A new demo notebook for a simple clustering model--->
  32: - Addition of new tests, `ClusterPerformanceMetrics`, `ClusterSizeDistribution`, `SilhouettePlot`, `HyperParametersTuning`, and `KMeansClustersOptimization`, to support clustering model evaluation 
  33  

site/releases/2023/2023-oct-25/highlights.qmd:
   91  <!---[SC-2346] Rouge and Bert score tests should show average scores by @juanmleng in [#263](https://github.com/validmind/developer-framework/pull/263)--->
   92: - **Rouge and Bert Score tests now show average scores**: Introduced `RougeMetricsAggregate` and `BertScoreAggregate` to offer a high-level overview of model performance across a large number of text rows. These tests complement the detailed row-by-row analysis provided by `RougeMetrics` and `BertScore`.
   93  <!--- NR this notebook is not currently included in our docs site:
   94: Tested these metrics running `foundational_models_summarization_high_code.ipynb` --->
   95  

  103  <!--- NR this notebook is not currently included in our docs site:
  104: To test these metrics, see notebook `foundation_models_summarization_bias.ipynb`---> 
  105  

site/releases/2023/2023-sep-27/highlights.qmd:
   20  <!---John6797/sc 2088/implement prompt validation tests poc by @johnwalz97 in [#232](https://github.com/validmind/validmind-python/pull/232)--->
   21: - **Prompt validation demo notebook for LLMs**. As a proof of concept, we added initial native prompt validation tests to the {{< var vm.developer >}}, including a notebook and simple template to test out these metrics on a sentiment analysis LLM model we built.
   22  

   24  <!---[SC-2144] Support hugging face summarization models by @AnilSorathiya in [#238](https://github.com/validmind/validmind-python/pull/238)--->
   25: - **Text summarization model demo notebook for LLMs**. We added a new notebook in the {{< var vm.developer >}} that includes the financial news dataset, initializes a Hugging Face summarization model using the `init_model` interface, implements relevant metrics for testing, and demonstrates how to run a _text summarization metrics_ test suite for an LLM instructed as a financial news summarizer.
   26  

  115  
  116: <!---[SC-2108] Multi-class tests should report metrics for individual classes and many other changes by @cachafla in [#228](https://github.com/validmind/validmind-python/pull/228)--->
  117  

  126  <!---SC-2100: Implement classification model comparison tests by @AnilSorathiya in [#231](https://github.com/validmind/validmind-python/pull/231)--->
  127: - **Implemented classification model comparison tests**. Added a model performance comparison test for classification tasks. The test includes metrics such as accuracy, F1, precision, recall, and roc_auc score.
  128  

site/releases/2024-feb-14/highlights.qmd:
  15  
  16: Documentation templates have been updated to support logging each test run as a unique result, making it possible to run the same test across different datasets or models. To make use of this new feature, you simply add a unique `result_id` identifier as a suffix to a `content_id` identifier in the content block definition of a `metric` or `test` content type. 
  17  

  22    content_id: validmind.data_validation.Skewness:training_data
  23: - content_type: metric
  24    content_id: validmind.data_validation.Skewness:test_data

site/releases/2024-mar-27/highlights.qmd:
  145  
  146: <!---Apply changes from juan5508/sc-3259/Create-Metric-V2-class by @nrichers in [#1](https://github.com/validmind/developer-framework/pull/1)--->
  147  <!--- NR Commented out as per Slack convo with Andres
  148: Apply changes from juan5508/sc-3259/Create-Metric-V2-class
  149: : This pull request introduces the Unit Metric feature, a new single value measure designed to streamline the risk assessment process in Machine Learning and AI model development. The Unit Metric simplifies complex evidence into actionable insights, allowing for efficient monitoring and comparison of risks across different models or datasets.
  150  

  152  ```python
  153: metric_id = "validmind.unit_metrics.sklearn.classification.F1"
  154  

  160  result = vm.unit_metrics.run_metric(
  161:     metric_id=metric_id, 
  162      inputs=inputs,

  176  **Changes**
  177: - Implemented the `UnitMetric` class.
  178: - Implemented sklearn classification metrics as unit metrics: `F1`, `Accuracy`, `Precision`, `Recall` and `ROC_AUC`
  179: - Added `vm.unit_metrics.run_metric` to the ValidMind client library to support running metrics. 
  180  
  181  **Testing**
  182: - Provided a notebook on defining, executing, and extracting results from Unit Metrics using a customer churn use case. See `run_unit_metrics.ipynb`
  183  --->

site/releases/2024-may-22/release-notes.qmd:
  129  
  130: We introduced a new `metric` decorator that turns any function into a ValidMind Test that you can use in your documentation. To learn what this decorator can do for you, try our code sample on JupyterHub!
  131  

  349  
  350: You can now register arbitrary extra columns in a dataset when a test needs to compute metrics outside of the existing sets of columns (features, targets, predictions). 
  351  
  352: For example, credit risk-related metrics may require access to a list of scores computed from predictions. In this case, an extra column called `scores` is needed for computing the metrics.
  353  

  494  <!---
  495: PR #48: John6797/sc 3881/support for metadata in new metric decorator
  496  URL: https://github.com/validmind/developer-framework/pull/48

  498  --->
  499: ### Support for metadata in new metric decorator
  500  
  501: We added new decorators to support task type and tag metadata in functional metrics.
  502  

  505  ```python
  506: from sklearn.metrics import accuracy_score
  507  
```

### CODE REVIEW

**Review by nrichers — COMMENTED** [review]
These changes generally look safe and help our docs catch up with a change that was introduced some time ago. 🍷 🍕 

I'd approve this EXCEPT it feels like this template schema has got to be staler than the stalest bread this side of Alpha Centauri: 

image

@cachafla has this template schema changed at all? IIRC, I pasted this into the docs sometime last year and it's not been updated since. Is this still current?

**Comment by validbeck** [comment]
> I'd approve this EXCEPT it feels like this template schema has got to be staler than the stalest bread this side of Alpha Centauri:

@nrichers I agree, but I think that is out of the scope of this PR (in the philosophy of your "atomic" PRs which I wholeheartedly endorse!). Can we compromise and approve this PR, and I will make a related Story to deal with the potentially outdated JSON schema, especially as I only half half an idea of how it works and would need to dive into it a bit more before I make changes?

**Comment by nrichers** [comment]
> @nrichers I agree, but I think that is out of the scope of this PR (in the philosophy of your "atomic" PRs which I wholeheartedly endorse!). Can we compromise and approve this PR, and I will make a related Story to deal with the potentially outdated JSON schema, especially as I only have half an idea of how it works and would need to dive into it a bit more before I make changes?

Yes, that's fair! I mainly asked because once of the bits of content you updated affects the template schemas, but let's sort that out separately.

**Review by nrichers — APPROVED** [review]
Content changes in this PR LGTM 🚀 but we should follow up with @cachafla to see if the template schema needs updating separately. 

FWIW, I checked the GitHub history for the template schema docs. They predate me squashing the history in preparation for making the documentation repo public — the oldest update I can find dates back to early July 2023. So these docs are very, very old.
